File: PythonAPI/carla/agents/navigation/local_planner.py
# Copyright (c) # Copyright (c) 2018-2020 CVC.
#
# This work is licensed under the terms of the MIT license.
# For a copy, see <https://opensource.org/licenses/MIT>.

""" This module contains a local planner to perform low-level waypoint following based on PID controllers. """

from enum import Enum
from collections import deque
import random

import carla
from agents.navigation.controller import VehiclePIDController
from agents.tools.misc import draw_waypoints


class RoadOption(Enum):
    """
    RoadOption represents the possible topological configurations when moving from a segment of lane to other.
    """
    VOID = -1
    LEFT = 1
    RIGHT = 2
    STRAIGHT = 3
    LANEFOLLOW = 4
    CHANGELANELEFT = 5
    CHANGELANERIGHT = 6


class LocalPlanner(object):
    """
    LocalPlanner implements the basic behavior of following a trajectory of waypoints that is generated on-the-fly.
    The low-level motion of the vehicle is computed by using two PID controllers, one is used for the lateral control
    and the other for the longitudinal control (cruise speed).

    When multiple paths are available (intersections) this local planner makes a random choice.
    """

    # minimum distance to target waypoint as a percentage (e.g. within 90% of
    # total distance)
    MIN_DISTANCE_PERCENTAGE = 0.9

    def __init__(self, vehicle, opt_dict=None):
        """
        :param vehicle: actor to apply to local planner logic onto
        :param opt_dict: dictionary of arguments with the following semantics:
            dt -- time difference between physics control in seconds. This is typically fixed from server side
                  using the arguments -benchmark -fps=F . In this case dt = 1/F

            target_speed -- desired cruise speed in Km/h

            sampling_radius -- search radius for next waypoints in seconds: e.g. 0.5 seconds ahead

            lateral_control_dict -- dictionary of arguments to setup the lateral PID controller
                                    {'K_P':, 'K_D':, 'K_I':, 'dt'}

            longitudinal_control_dict -- dictionary of arguments to setup the longitudinal PID controller
                                        {'K_P':, 'K_D':, 'K_I':, 'dt'}
        """
        self._vehicle = vehicle
        self._map = self._vehicle.get_world().get_map()

        self._dt = None
        self._target_speed = None
        self._sampling_radius = None
        self._min_distance = None
        self._current_waypoint = None
        self._target_road_option = None
        self._next_waypoints = None
        self.target_waypoint = None
        self._vehicle_controller = None
        self._stop_waypoint_creation = None
        # queue with tuples of (waypoint, RoadOption)
        self._waypoints_queue = deque(maxlen=20000)
        self._buffer_size = 5
        self._waypoint_buffer = deque(maxlen=self._buffer_size)

        # initializing controller
        self._init_controller(opt_dict)

    def __del__(self):
        if self._vehicle:
            self._vehicle.destroy()
            print("Destroying ego-vehicle!")

    def reset_vehicle(self):
        self._vehicle = None
        print("Resetting ego-vehicle!")

    def _init_controller(self, opt_dict):
        """
        Controller initialization.

        :param opt_dict: dictionary of arguments.
        :return:
        """
        # default params
        self._dt = 1.0 / 20.0
        self._target_speed = 20.0  # Km/h
        self._sampling_radius = self._target_speed * 1 / 3.6  # 1 seconds horizon
        self._min_distance = self._sampling_radius * self.MIN_DISTANCE_PERCENTAGE
        self._max_brake = 0.3
        self._max_throt = 0.75
        self._max_steer = 0.8
        args_lateral_dict = {
            'K_P': 1.95,
            'K_D': 0.2,
            'K_I': 0.07,
            'dt': self._dt}
        args_longitudinal_dict = {
            'K_P': 1.0,
            'K_D': 0,
            'K_I': 0.05,
            'dt': self._dt}
        self._offset = 0

        # parameters overload
        if opt_dict:
            if 'dt' in opt_dict:
                self._dt = opt_dict['dt']
            if 'target_speed' in opt_dict:
                self._target_speed = opt_dict['target_speed']
            if 'sampling_radius' in opt_dict:
                self._sampling_radius = self._target_speed * \
                                        opt_dict['sampling_radius'] / 3.6
            if 'lateral_control_dict' in opt_dict:
                args_lateral_dict = opt_dict['lateral_control_dict']
            if 'longitudinal_control_dict' in opt_dict:
                args_longitudinal_dict = opt_dict['longitudinal_control_dict']
            if 'max_throttle' in opt_dict:
                self._max_throt = opt_dict['max_throttle']
            if 'max_brake' in opt_dict:
                self._max_brake = opt_dict['max_brake']
            if 'max_steering' in opt_dict:
                self._max_steer = opt_dict['max_steering']
            if 'offset' in opt_dict:
                self._offset = opt_dict['offset']

        self._current_waypoint = self._map.get_waypoint(self._vehicle.get_location())
        self._vehicle_controller = VehiclePIDController(self._vehicle,
                                                        args_lateral=args_lateral_dict,
                                                        args_longitudinal=args_longitudinal_dict,
                                                        offset=self._offset,
                                                        max_throttle=self._max_throt,
                                                        max_brake=self._max_brake,
                                                        max_steering=self._max_steer)

        self._stop_waypoint_creation = False

        # compute initial waypoints
        self._waypoints_queue.append((self._current_waypoint.next(self._sampling_radius)[0], RoadOption.LANEFOLLOW))

        self._target_road_option = RoadOption.LANEFOLLOW
        # fill waypoint trajectory queue
        self._compute_next_waypoints(k=200)

    def set_speed(self, speed):
        """
        Request new target speed.

        :param speed: new target speed in Km/h
        :return:
        """
        self._target_speed = speed

    def _compute_next_waypoints(self, k=1):
        """
        Add new waypoints to the trajectory queue.

        :param k: how many waypoints to compute
        :return:
        """
        # check we do not overflow the queue
        available_entries = self._waypoints_queue.maxlen - len(self._waypoints_queue)
        k = min(available_entries, k)

        for _ in range(k):
            last_waypoint = self._waypoints_queue[-1][0]
            next_waypoints = list(last_waypoint.next(self._sampling_radius))

            if len(next_waypoints) == 0:
                break
            elif len(next_waypoints) == 1:
                # only one option available ==> lanefollowing
                next_waypoint = next_waypoints[0]
                road_option = RoadOption.LANEFOLLOW
            else:
                # random choice between the possible options
                road_options_list = _retrieve_options(
                    next_waypoints, last_waypoint)
                road_option = random.choice(road_options_list)
                next_waypoint = next_waypoints[road_options_list.index(
                    road_option)]

            self._waypoints_queue.append((next_waypoint, road_option))

    def set_global_plan(self, current_plan, stop_waypoint_creation=True):
        """
        Resets the waypoint queue and buffer to match the new plan. Also
        handles the stop_waypoint_creation flag to avoid creating more waypoints

        :param current_plan: list of (carla.Waypoint, RoadOption)
        :param stop_waypoint_creation: bool
        :return:
        """

        # Reset the queue
        self._waypoints_queue.clear()
        for elem in current_plan:
            self._waypoints_queue.append(elem)
        self._target_road_option = RoadOption.LANEFOLLOW

        # and the buffer
        self._waypoint_buffer.clear()
        for _ in range(self._buffer_size):
            if self._waypoints_queue:
                self._waypoint_buffer.append(
                    self._waypoints_queue.popleft())
            else:
                break

        self._stop_waypoint_creation = stop_waypoint_creation

    def run_step(self, debug=False):
        """
        Execute one step of local planning which involves running the longitudinal and lateral PID controllers to
        follow the waypoints trajectory.

        :param debug: boolean flag to activate waypoints debugging
        :return: control to be applied
        """

        # not enough waypoints in the horizon? => add more!
        if not self._stop_waypoint_creation and len(self._waypoints_queue) < int(self._waypoints_queue.maxlen * 0.5):
            self._compute_next_waypoints(k=100)

        if len(self._waypoints_queue) == 0 and len(self._waypoint_buffer) == 0:
            control = carla.VehicleControl()
            control.steer = 0.0
            control.throttle = 0.0
            control.brake = 1.0
            control.hand_brake = False
            control.manual_gear_shift = False

            return control

        #   Buffering the waypoints
        if not self._waypoint_buffer:
            for _ in range(self._buffer_size):
                if self._waypoints_queue:
                    self._waypoint_buffer.append(
                        self._waypoints_queue.popleft())
                else:
                    break

        # current vehicle waypoint
        vehicle_transform = self._vehicle.get_transform()
        self._current_waypoint = self._map.get_waypoint(vehicle_transform.location)
        # target waypoint
        self.target_waypoint, self._target_road_option = self._waypoint_buffer[0]
        # move using PID controllers
        control = self._vehicle_controller.run_step(self._target_speed, self.target_waypoint)

        # purge the queue of obsolete waypoints
        max_index = -1

        for i, (waypoint, _) in enumerate(self._waypoint_buffer):
            if waypoint.transform.location.distance(vehicle_transform.location) < self._min_distance:
                max_index = i
        if max_index >= 0:
            for i in range(max_index + 1):
                self._waypoint_buffer.popleft()

        if debug:
            draw_waypoints(self._vehicle.get_world(), [self.target_waypoint], self._vehicle.get_location().z + 1.0)

        return control

    def done(self):
        """
        Returns whether or not the planner has finished

        :return: boolean
        """
        return len(self._waypoints_queue) == 0 and len(self._waypoint_buffer) == 0

def _retrieve_options(list_waypoints, current_waypoint):
    """
    Compute the type of connection between the current active waypoint and the multiple waypoints present in
    list_waypoints. The result is encoded as a list of RoadOption enums.

    :param list_waypoints: list with the possible target waypoints in case of multiple options
    :param current_waypoint: current active waypoint
    :return: list of RoadOption enums representing the type of connection from the active waypoint to each
             candidate in list_waypoints
    """
    options = []
    for next_waypoint in list_waypoints:
        # this is needed because something we are linking to
        # the beggining of an intersection, therefore the
        # variation in angle is small
        next_next_waypoint = next_waypoint.next(3.0)[0]
        link = _compute_connection(current_waypoint, next_next_waypoint)
        options.append(link)

    return options


def _compute_connection(current_waypoint, next_waypoint, threshold=35):
    """
    Compute the type of topological connection between an active waypoint (current_waypoint) and a target waypoint
    (next_waypoint).

    :param current_waypoint: active waypoint
    :param next_waypoint: target waypoint
    :return: the type of topological connection encoded as a RoadOption enum:
             RoadOption.STRAIGHT
             RoadOption.LEFT
             RoadOption.RIGHT
    """
    n = next_waypoint.transform.rotation.yaw
    n = n % 360.0

    c = current_waypoint.transform.rotation.yaw
    c = c % 360.0

    diff_angle = (n - c) % 180.0
    if diff_angle < threshold or diff_angle > (180 - threshold):
        return RoadOption.STRAIGHT
    elif diff_angle > 90.0:
        return RoadOption.LEFT
    else:
        return RoadOption.RIGHT

File: LibCarla/source/carla/agents/navigation/LocalPlanner.cpp
// Copyright (c) 2018-2020 CVC.
//
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT>.

#include "carla/agents/navigation/LocalPlanner.h"

#include "carla/agents/navigation/Misc.h"
#include "carla/client/Map.h"
#include "carla/client/Vehicle.h"
#include "carla/client/Waypoint.h"
#include "carla/client/World.h"
#include "carla/geom/Location.h"
#include "carla/geom/Transform.h"

#include <algorithm>
#include <cmath>
#include <iostream>

namespace carla {
namespace agents {
namespace navigation {

namespace {

constexpr float kKmhToMs = 1.0f / 3.6f;

/// Compute the topological connection between @a current and @a next_wp,
/// matching `_compute_connection` in local_planner.py.
RoadOption ComputeConnection(
    SharedPtr<client::Waypoint> current,
    SharedPtr<client::Waypoint> next_wp,
    float threshold = 35.0f) {
  // Python uses `% 360` and `% 180` on yaws to wrap them. Replicate with
  // `std::fmod` and a positive-fixup, since C++ `%` doesn't accept floats.
  auto wrap = [](float v, float mod) {
    float r = std::fmod(v, mod);
    if (r < 0.0f) r += mod;
    return r;
  };

  const float n = wrap(next_wp->GetTransform().rotation.yaw, 360.0f);
  const float c = wrap(current->GetTransform().rotation.yaw, 360.0f);
  const float diff = wrap(n - c, 180.0f);

  if (diff < threshold || diff > (180.0f - threshold)) {
    return RoadOption::Straight;
  }
  if (diff > 90.0f) {
    return RoadOption::Left;
  }
  return RoadOption::Right;
}

/// Mirror of `_retrieve_options` in local_planner.py.
std::vector<RoadOption> RetrieveOptions(
    const std::vector<SharedPtr<client::Waypoint>> &candidates,
    SharedPtr<client::Waypoint> current) {
  std::vector<RoadOption> options;
  options.reserve(candidates.size());
  for (const auto &next_wp : candidates) {
    // The Python code probes 3 m further to disambiguate intersections that
    // share an entry waypoint.
    auto next_next = next_wp->GetNext(3.0);
    if (next_next.empty()) {
      options.push_back(ComputeConnection(current, next_wp));
    } else {
      options.push_back(ComputeConnection(current, next_next.front()));
    }
  }
  return options;
}

} // namespace

// =========================================================================
// LocalPlanner
// =========================================================================

LocalPlanner::LocalPlanner(
    SharedPtr<client::Vehicle> vehicle,
    const Options &opt,
    SharedPtr<client::Map> map)
  : _vehicle(std::move(vehicle)),
    _map(std::move(map)),
    _dt(opt.dt),
    _target_speed(opt.target_speed),
    _sampling_radius(opt.sampling_radius),
    _lateral_args(opt.lateral),
    _longitudinal_args(opt.longitudinal),
    _max_throttle(opt.max_throttle),
    _max_brake(opt.max_brake),
    _max_steering(opt.max_steering),
    _offset(opt.offset),
    _base_min_distance(opt.base_min_distance),
    _distance_ratio(opt.distance_ratio),
    _follow_speed_limits(opt.follow_speed_limits),
    _rng(std::random_device{}()) {
  // If no map was supplied, pull it from the vehicle's world (mirrors the
  // `map_inst` fallback in local_planner.py).
  if (!_map && _vehicle) {
    _map = _vehicle->GetWorld().GetMap();
  }
  InitController();
}

void LocalPlanner::InitController() {
  _vehicle_controller = std::make_unique<VehiclePIDController>(
      _vehicle,
      _lateral_args,
      _longitudinal_args,
      _offset,
      _max_throttle,
      _max_brake,
      _max_steering);

  if (_vehicle && _map) {
    auto current_waypoint = _map->GetWaypoint(_vehicle->GetLocation());
    _target_waypoint = current_waypoint;
    _target_road_option = RoadOption::LaneFollow;
    _waypoints_queue.emplace_back(_target_waypoint, _target_road_option);
  }
}

void LocalPlanner::ResetVehicle() {
  _vehicle.reset();
}

void LocalPlanner::SetSpeed(float speed_kmh) {
  if (_follow_speed_limits) {
    std::cout << "WARNING: The max speed is currently set to follow the speed"
                 " limits. Use 'follow_speed_limits' to deactivate this\n";
  }
  _target_speed = speed_kmh;
}

void LocalPlanner::FollowSpeedLimits(bool value) {
  _follow_speed_limits = value;
}

void LocalPlanner::ComputeNextWaypoints(std::size_t k) {
  // Bound k by remaining queue capacity, like Python's
  // `available_entries = self._waypoints_queue.maxlen - len(...)`.
  const std::size_t available = (_waypoints_queue.size() < _max_queue_size)
      ? (_max_queue_size - _waypoints_queue.size())
      : 0;
  k = std::min(k, available);

  for (std::size_t i = 0; i < k; ++i) {
    if (_waypoints_queue.empty()) break;
    auto last_waypoint = _waypoints_queue.back().first;
    if (!last_waypoint) break;
    auto next_waypoints = last_waypoint->GetNext(_sampling_radius);

    if (next_waypoints.empty()) {
      break;
    }

    SharedPtr<client::Waypoint> next_waypoint;
    RoadOption road_option;
    if (next_waypoints.size() == 1) {
      next_waypoint = next_waypoints.front();
      road_option = RoadOption::LaneFollow;
    } else {
      // Random choice among the available branches.
      auto road_options_list = RetrieveOptions(next_waypoints, last_waypoint);
      std::uniform_int_distribution<std::size_t> dist(
          0, road_options_list.size() - 1);
      const std::size_t idx = dist(_rng);
      road_option = road_options_list[idx];
      // Python pulls `next_waypoints[road_options_list.index(road_option)]`,
      // i.e. the *first* candidate that shares the chosen RoadOption. With a
      // pre-randomised idx that's just the same as picking idx directly,
      // because there's a 1:1 alignment between the lists (the first option
      // matching the random pick is the random pick itself).
      next_waypoint = next_waypoints[idx];
    }

    _waypoints_queue.emplace_back(next_waypoint, road_option);
  }
}

void LocalPlanner::SetGlobalPlan(
    const std::vector<PlanItem> &plan,
    bool stop_waypoint_creation,
    bool clean_queue) {
  if (clean_queue) {
    _waypoints_queue.clear();
  }

  // The Python implementation grows the deque's `maxlen` whenever the new
  // plan exceeds it; we keep a soft cap and just raise it as needed.
  const std::size_t new_plan_length =
      plan.size() + _waypoints_queue.size();
  if (new_plan_length > _max_queue_size) {
    _max_queue_size = new_plan_length;
  }

  for (const auto &elem : plan) {
    _waypoints_queue.push_back(elem);
  }
  _stop_waypoint_creation = stop_waypoint_creation;
}

void LocalPlanner::SetOffset(float offset) {
  _offset = offset;
  if (_vehicle_controller) {
    _vehicle_controller->SetOffset(offset);
  }
}

rpc::VehicleControl LocalPlanner::RunStep(bool debug) {
  if (_follow_speed_limits && _vehicle) {
    _target_speed = _vehicle->GetSpeedLimit();
  }

  if (!_stop_waypoint_creation
      && _waypoints_queue.size() < _min_queue_length) {
    ComputeNextWaypoints(_min_queue_length);
  }

  // Purge waypoints we've already passed.
  if (_vehicle) {
    const auto veh_location = _vehicle->GetLocation();
    const float vehicle_speed_ms = GetSpeed(_vehicle) * kKmhToMs;
    const float min_distance =
        _base_min_distance + _distance_ratio * vehicle_speed_ms;

    std::size_t num_removed = 0;
    for (const auto &item : _waypoints_queue) {
      // Don't remove the very last waypoint until we're almost on top of it
      // (mirrors Python's `min_distance = 1` in the same situation).
      const float threshold =
          (_waypoints_queue.size() - num_removed == 1) ? 1.0f : min_distance;

      const auto &wp_loc = item.first->GetTransform().location;
      if (veh_location.Distance(wp_loc) < threshold) {
        ++num_removed;
      } else {
        break;
      }
    }
    for (std::size_t i = 0; i < num_removed; ++i) {
      _waypoints_queue.pop_front();
    }
  }

  rpc::VehicleControl control;
  if (_waypoints_queue.empty()) {
    control.steer = 0.0f;
    control.throttle = 0.0f;
    control.brake = 1.0f;
    control.hand_brake = false;
    control.manual_gear_shift = false;
  } else {
    _target_waypoint = _waypoints_queue.front().first;
    _target_road_option = _waypoints_queue.front().second;
    control = _vehicle_controller->RunStep(_target_speed, _target_waypoint);
  }

  if (debug && _vehicle && _target_waypoint) {
    auto world = std::make_shared<client::World>(_vehicle->GetWorld());
    DrawWaypoints(world, {_target_waypoint}, 1.0f);
  }

  return control;
}

std::pair<SharedPtr<client::Waypoint>, RoadOption>
LocalPlanner::GetIncomingWaypointAndDirection(int steps) const {
  if (steps >= 0
      && static_cast<std::size_t>(steps) < _waypoints_queue.size()) {
    return _waypoints_queue[static_cast<std::size_t>(steps)];
  }
  if (!_waypoints_queue.empty()) {
    return _waypoints_queue.back();
  }
  return {nullptr, RoadOption::Void};
}

const std::deque<LocalPlanner::PlanItem> &LocalPlanner::GetPlan() const {
  return _waypoints_queue;
}

bool LocalPlanner::Done() const {
  return _waypoints_queue.empty();
}

SharedPtr<client::Waypoint> LocalPlanner::TargetWaypoint() const {
  return _target_waypoint;
}

RoadOption LocalPlanner::TargetRoadOption() const {
  return _target_road_option;
}

} // namespace navigation
} // namespace agents
} // namespace carla
